rules/key-principles-of-api-design/rule.md

```diff
@@ -0,0 +1,80 @@
+
+Building a high-quality API means adhering to best practices that enhance clarity, reliability, and security, creating a more consistent experience for developers and users.
+
+<!--endintro-->
+
+`youtube: https://youtu.be/_gQaygjm_hg`
+**Video: Good APIs Vs Bad APIs: 7 Tips for API Design (6 min)**
+
+## Tip \#1 - Use clear naming  
+
+Choose descriptive, intuitive names for endpoints and parameters to improve readability. Adhere to REST conventions, using nouns and verbs logically (e.g., `/users` for accessing user data, `/users/{id}` for specific user information).
+
+Clear naming simplifies usage and reduces misunderstandings.  
+
+## Tip \#2 - Ensure reliability through idempotency  
+
+To avoid unintended actions from repeated requests, design POST, PUT, and DELETE operations as idempotent, where repeating an action yields the same result as performing it once.
+
+For example, if a DELETE request removes a record, re-sending it should not throw errors if the record is already deleted.
+
+This prevents accidental duplicate data processing.  
+
+## Tip \#3 - Implement versioning  
+
+Introduce versioning from the start (e.g., `/v1/resource`) to maintain backward compatibility when updating the API.
+
+Versioning helps users manage changes without breaking existing implementations, allowing them to adopt new features gradually.  
+
+## Tip \#4 - Add pagination for responses  
+
+For endpoints that return lists, apply pagination to prevent overwhelming the client with too much data.
+
+Use query parameters like ?page= and ?limit= to specify page numbers and size, offering a more manageable data experience while improving performance.  
+
+## Tip \#5 - Use clear query strings for sorting  
+
+When supporting sorting, apply clear query strings (e.g., `?sortBy=name&order=asc`).
+
+Consistent sorting parameters allow developers to retrieve and organize data efficiently and minimize confusion in handling API responses.  
+
+## Tip \#6 - Security should not be an afterthought
+
+Security is paramount in API design. Pass sensitive information, like access tokens, securely via headers (e.g., `Authorization: Bearer {token}`).
+
+Enforce HTTPS for encrypted communication and consider OAuth for user authentication and authorization, protecting against unauthorized access.  
+
+## Tip \#7 - Keep cross-resource references simple
+
+For APIs that reference multiple resources (e.g., `userId` in a post endpoint), keep relationships simple to prevent over-complicating endpoints.
+
+Provide clear references or IDs rather than nested data whenever possible to keep API responses readable and easy to follow.  
+
+## Extra tip - Rate limiting  
+
+Implement rate limiting to control the number of requests per user within a time frame, protecting the API from abuse and ensuring fair resource usage.
+
+Provide appropriate status codes and messages (e.g., `HTTP 429`) when limits are reached to inform users clearly.  
+
```

Comment:
**Extra tip - Caching**

Add this as an extra section

Reply:
As per our conversation, I will also add an extra section about **Compression**

Reply:
Done - I've added both sections